File: apps/android/app/src/main/java/it/mtre_consulting/chroma/viewmodel/AppViewModel.kt
package it.mtre_consulting.chroma.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import it.mtre_consulting.chroma.data.model.Colour
import it.mtre_consulting.chroma.data.model.Palette
import it.mtre_consulting.chroma.data.model.Token
import it.mtre_consulting.chroma.data.model.TokenGroup
import it.mtre_consulting.chroma.data.model.TokenValue
import it.mtre_consulting.chroma.data.store.AppDataStore
import it.mtre_consulting.chroma.util.generateId
import it.mtre_consulting.chroma.util.hexToColour
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AppViewModel(private val dataStore: AppDataStore) : ViewModel() {

    private val _palettes = MutableStateFlow<List<Palette>>(emptyList())
    val palettes: StateFlow<List<Palette>> = _palettes.asStateFlow()

    private val _tokenGroups = MutableStateFlow<List<TokenGroup>>(emptyList())
    val tokenGroups: StateFlow<List<TokenGroup>> = _tokenGroups.asStateFlow()

    init {
        viewModelScope.launch {
            dataStore.palettesFlow.collect { _palettes.value = it }
        }
        viewModelScope.launch {
            dataStore.tokenGroupsFlow.collect { _tokenGroups.value = it }
        }
    }

    // ── Palettes ────────────────────────────────────────────────────────────

    fun addPalette(name: String) = mutatePalettes { palettes ->
        palettes + Palette(id = generateId(), name = name)
    }

    fun removePalette(id: String) = mutatePalettes { it.filter { p -> p.id != id } }

    fun addColour(paletteId: String, hex: String, name: String = "") = mutatePalettes { palettes ->
        palettes.map { p ->
            if (p.id != paletteId) p
            else {
                val colour = hexToColour(hex, name.ifBlank { "Untitled" }, id = generateId())
                p.copy(colours = p.colours + colour, updatedAt = System.currentTimeMillis())
            }
        }
    }

    fun removeColour(paletteId: String, colourId: String) = mutatePalettes { palettes ->
        palettes.map { p ->
            if (p.id != paletteId) p
            else p.copy(
                colours = p.colours.filter { it.id != colourId },
                updatedAt = System.currentTimeMillis(),
            )
        }
    }

    fun updateColour(paletteId: String, colourId: String, patch: Colour) =
        mutatePalettes { palettes ->
            palettes.map { p ->
                if (p.id != paletteId) p
                else p.copy(
                    colours = p.colours.map { if (it.id == colourId) patch else it },
                    updatedAt = System.currentTimeMillis(),
                )
            }
        }

    // ── Token Groups ────────────────────────────────────────────────────────

    fun addGroup(name: String) = mutateGroups { groups ->
        groups + TokenGroup(id = generateId(), name = name)
    }

    fun removeGroup(groupId: String) = mutateGroups { it.filter { g -> g.id != groupId } }

    fun renameGroup(groupId: String, name: String) = mutateGroups { groups ->
        groups.map { if (it.id == groupId) it.copy(name = name) else it }
    }

    fun addToken(groupId: String, name: String) = mutateGroups { groups ->
        groups.map { g ->
            if (g.id != groupId) g
            else g.copy(tokens = g.tokens + Token(id = generateId(), name = name))
        }
    }

    fun removeToken(groupId: String, tokenId: String) = mutateGroups { groups ->
        groups.map { g ->
            if (g.id != groupId) g
            else g.copy(tokens = g.tokens.filter { it.id != tokenId })
        }
    }

    fun assignColour(groupId: String, tokenId: String, paletteId: String, colourId: String) =
        mutateGroups { groups ->
            groups.map { g ->
                if (g.id != groupId) g
                else g.copy(
                    tokens = g.tokens.map { t ->
                        if (t.id != tokenId) t
                        else t.copy(value = TokenValue(colourId = colourId, paletteId = paletteId))
                    }
                )
            }
        }

    // ── Helpers ─────────────────────────────────────────────────────────────

    private fun mutatePalettes(transform: (List<Palette>) -> List<Palette>) {
        val updated = transform(_palettes.value)
        _palettes.value = updated
        viewModelScope.launch { dataStore.savePalettes(updated) }
    }

    private fun mutateGroups(transform: (List<TokenGroup>) -> List<TokenGroup>) {
        val updated = transform(_tokenGroups.value)
        _tokenGroups.value = updated
        viewModelScope.launch { dataStore.saveTokenGroups(updated) }
    }
}

class AppViewModelFactory(private val dataStore: AppDataStore) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return AppViewModel(dataStore) as T
    }
}

File: apps/ios/Chroma/Data/AppViewModel.swift
//
//  AppViewModel.swift
//  Chroma
//
//  Created by Simone Rolando (M-Tre Consulting) on 02/05/2026.
//

import Foundation
import Observation

/// Central view model for the app. Owns all in-memory state and delegates
/// persistence to ``DataStore``.
///
/// Inject a single instance at the root via `.environment()` and read it
/// in any view with `@Environment(AppViewModel.self)`.
@Observable
final class AppViewModel {

    /// All palettes created by the user.
    private(set) var palettes: [Palette] = []

    /// All token groups created by the user.
    private(set) var tokenGroups: [TokenGroup] = []

    private let store: DataStore

    init(store: DataStore) {
        self.store = store
        store.load()
        palettes = store.palettes
        tokenGroups = store.tokenGroups
    }

    // MARK: - Palettes

    /// Creates a new palette with the given name and appends it to the list.
    func addPalette(name: String) {
        mutatePalettes {
            $0.append(Palette(id: generateId(), name: name, colours: [], createdAt: nowMs, updatedAt: nowMs))
        }
    }

    /// Removes the palette with the given ID, along with all its colours.
    func removePalette(id: String) {
        mutatePalettes { $0.removeAll { $0.id == id } }
    }

    /// Parses `hex` into a ``Colour`` and appends it to the specified palette.
    func addColour(paletteId: String, hex: String, name: String = "") {
        mutatePalettes { palettes in
            guard let i = palettes.firstIndex(where: { $0.id == paletteId }) else { return }
            let colour = hexToColour(hex: hex, name: name.isEmpty ? "Untitled" : name, id: generateId())
            palettes[i].colours.append(colour)
            palettes[i].updatedAt = nowMs
        }
    }

    /// Removes a colour from the specified palette.
    func removeColour(paletteId: String, colourId: String) {
        mutatePalettes { palettes in
            guard let i = palettes.firstIndex(where: { $0.id == paletteId }) else { return }
            palettes[i].colours.removeAll { $0.id == colourId }
            palettes[i].updatedAt = nowMs
        }
    }

    /// Replaces a colour in the specified palette with `patch`.
    func updateColour(paletteId: String, colourId: String, patch: Colour) {
        mutatePalettes { palettes in
            guard let pi = palettes.firstIndex(where: { $0.id == paletteId }),
                  let ci = palettes[pi].colours.firstIndex(where: { $0.id == colourId })
            else { return }
            palettes[pi].colours[ci] = patch
            palettes[pi].updatedAt = nowMs
        }
    }

    // MARK: - Token Groups

    /// Creates a new empty token group with the given name.
    func addGroup(name: String) {
        mutateGroups {
            $0.append(TokenGroup(id: generateId(), name: name, tokens: []))
        }
    }

    /// Removes the token group with the given ID and all its tokens.
    func removeGroup(groupId: String) {
        mutateGroups { $0.removeAll { $0.id == groupId } }
    }

    /// Renames an existing token group.
    func renameGroup(groupId: String, name: String) {
        mutateGroups { groups in
            guard let i = groups.firstIndex(where: { $0.id == groupId }) else { return }
            groups[i] = TokenGroup(id: groups[i].id, name: name, tokens: groups[i].tokens)
        }
    }

    /// Creates a new token with the given name inside the specified group.
    func addToken(groupId: String, name: String) {
        mutateGroups { groups in
            guard let i = groups.firstIndex(where: { $0.id == groupId }) else { return }
            groups[i].tokens.append(Token(id: generateId(), name: name, value: TokenValue(colourId: "", paletteId: "")))
        }
    }

    /// Removes a token from the specified group.
    func removeToken(groupId: String, tokenId: String) {
        mutateGroups { groups in
            guard let i = groups.firstIndex(where: { $0.id == groupId }) else { return }
            groups[i].tokens.removeAll { $0.id == tokenId }
        }
    }

    /// Binds a token to a specific colour within a palette.
    func assignColour(groupId: String, tokenId: String, paletteId: String, colourId: String) {
        mutateGroups { groups in
            guard let gi = groups.firstIndex(where: { $0.id == groupId }),
                  let ti = groups[gi].tokens.firstIndex(where: { $0.id == tokenId })
            else { return }
            groups[gi].tokens[ti].value = TokenValue(colourId: colourId, paletteId: paletteId)
        }
    }

    // MARK: - Helpers

    private var nowMs: Int64 { Int64(Date().timeIntervalSince1970 * 1000) }

    private func mutatePalettes(_ transform: (inout [Palette]) -> Void) {
        transform(&palettes)
        store.palettes = palettes
        store.savePalettes()
    }

    private func mutateGroups(_ transform: (inout [TokenGroup]) -> Void) {
        transform(&tokenGroups)
        store.tokenGroups = tokenGroups
        store.saveTokenGroups()
    }
}
